File: packages/flutter_amazon_connect_webrtc/ios/Classes/ChimeCallManager.swift
import AmazonChimeSDK
import AVFoundation
import Foundation

/// Receives call events from the manager. The plugin implements this and marshals every emission to
/// the main thread before pushing it onto the Flutter event channel (threading rule, spec §B.2).
protocol ChimeEventEmitter: AnyObject {
    func emit(_ event: [String: Any])
}

/// Owns the Amazon Chime `MeetingSession` lifecycle and the device audio session, and translates
/// Chime observer callbacks into the plugin's event contract.
final class ChimeCallManager: NSObject {
    private let logger = ConsoleLogger(name: "ChimeFlutter")
    private weak var emitter: ChimeEventEmitter?
    private(set) var meetingSession: MeetingSession?
    private var isMuted = false
    private var mediaStarted = false
    /// True once CallKit has activated the audio session. For an ANSWERED incoming call,
    /// `didActivate` fires before the join credentials are fetched — `join` checks this flag and
    /// starts media immediately instead of waiting for a `didActivate` that already happened.
    private var callKitAudioActive = false
    private var routeObserver: NSObjectProtocol?
    private var lastRoute: String?
    /// True after the user explicitly turned the speaker ON this call — stops the headset-default
    /// logic from fighting an intentional speaker choice.
    private var userForcedSpeaker = false

    init(emitter: ChimeEventEmitter) {
        self.emitter = emitter
        super.init()
    }

    // MARK: - Commands

    func join(sessionMap: [String: Any], callKitEnabled: Bool) throws {
        let configuration = try ChimeMeetingSessionAdapter.makeConfiguration(from: sessionMap)
        // With CallKit, the OS activates the audio session in the CXProvider's `didActivate`
        // delegate — we only set the category/mode here and must NOT call setActive ourselves.
        try configureAudioSession(activate: !callKitEnabled)

        let session = DefaultMeetingSession(configuration: configuration, logger: logger)
        meetingSession = session

        let audioVideo = session.audioVideo
        audioVideo.addAudioVideoObserver(observer: self)
        audioVideo.addRealtimeObserver(observer: self)
        audioVideo.addVideoTileObserver(observer: self)

        // With CallKit the session is not active yet — reading the route now would report the
        // ambient media default (loudspeaker), so the initial emit waits for activation.
        observeAudioRouteChanges(emitInitial: !callKitEnabled)
        emitState("connecting")

        if callKitEnabled {
            // CRITICAL: with CallKit, `audioVideo.start(callKitEnabled:)` MUST be called from the
            // CXProvider's `didActivate` (via startAudioVideoForCallKit()) — never here — or audio
            // will not start (per AWS's CallKit integration guidance). The one exception: an
            // answered INCOMING call whose didActivate already fired while the credentials were
            // being fetched — start now, the session is active.
            if callKitAudioActive { startAudioVideoForCallKit() }
            return
        }

        try audioVideo.start(callKitEnabled: false)
        // Remote video is opt-in on the Chime iOS SDK — without this the agent's video tile never
        // arrives (parity with Android). Harmless for audio-only calls.
        audioVideo.startRemoteVideo()
        applyPreferredHeadsetRoute()
    }

    /// Called by the CallKit coordinator from `CXProvider(_:didActivate:)`. Starts the Chime media
    /// once CallKit has activated the audio session (idempotent — didActivate may fire more than once).
    func startAudioVideoForCallKit() {
        callKitAudioActive = true
        guard let audioVideo = meetingSession?.audioVideo, !mediaStarted else { return }
        mediaStarted = true
        do {
            try audioVideo.start(callKitEnabled: true)
            audioVideo.startRemoteVideo()
            if isMuted { _ = audioVideo.realtimeLocalMute() }
            // The session is active now — route to a connected headset (phone-app default) and
            // report the real route (the initial emit was skipped for CallKit).
            applyPreferredHeadsetRoute()
            emitAudioRouteIfChanged()
        } catch {
            emitter?.emit(["type": "error", "code": "sdkError",
                           "message": "audio start failed: \(error)", "fatal": true])
            emitState("failed")
        }
    }

    func leave() {
        if let audioVideo = meetingSession?.audioVideo {
            audioVideo.stopLocalVideo()
            audioVideo.stopRemoteVideo()
            audioVideo.removeAudioVideoObserver(observer: self)
            audioVideo.removeRealtimeObserver(observer: self)
            audioVideo.removeVideoTileObserver(observer: self)
            audioVideo.stop()
        }
        meetingSession = nil
        mediaStarted = false
        isMuted = false
        callKitAudioActive = false
        userForcedSpeaker = false
        stopObservingAudioRouteChanges()
        deactivateAudioSession()
        emitState("disconnected")
    }

    /// CallKit deactivated the audio session (`didDeactivate`).
    func callKitAudioSessionEnded() {
        callKitAudioActive = false
    }

    @discardableResult
    func setMuted(_ muted: Bool) -> Bool {
        guard let audioVideo = meetingSession?.audioVideo else { return false }
        let ok = muted ? audioVideo.realtimeLocalMute() : audioVideo.realtimeLocalUnmute()
        if ok {
            isMuted = muted
            emitter?.emit(["type": "muteChanged", "muted": muted])
        }
        return ok
    }

    func setLocalVideoEnabled(_ enabled: Bool) throws {
        guard let audioVideo = meetingSession?.audioVideo else { return }
        if enabled {
            try audioVideo.startLocalVideo()
        } else {
            audioVideo.stopLocalVideo()
        }
    }

    func switchCamera() {
        meetingSession?.audioVideo.switchCamera()
    }

    func setSpeakerphone(_ enabled: Bool) {
        // Prefer the Chime SDK's device controller — it owns the audio session (with CallKit, a raw
        // AVAudioSession override can be reverted by the system). Fall back to the session override
        // if the expected device isn't listed. Speaker OFF returns to what the OS would pick —
        // bluetooth, then wired headset, then the earpiece — matching the system call screen.
        // The resulting route is reported via the routeChangeNotification observer, not assumed here.
        userForcedSpeaker = enabled
        if let audioVideo = meetingSession?.audioVideo {
            let devices = audioVideo.listAudioDevices()
            let preference: [MediaDeviceType] = enabled
                ? [.audioBuiltInSpeaker]
                : [.audioBluetooth, .audioWiredHeadset, .audioHandset]
            let target = preference.compactMap { wanted in devices.first { $0.type == wanted } }.first
            if let target = target {
                if !enabled {
                    // Chime's chooseAudioDevice(non-speaker) only sets the preferred INPUT — it
                    // never clears a loudspeaker override, so the output would stay on the speaker.
                    try? AVAudioSession.sharedInstance().overrideOutputAudioPort(.none)
                }
                audioVideo.chooseAudioDevice(mediaDevice: target)
                return
            }
        }
        let session = AVAudioSession.sharedInstance()
        try? session.overrideOutputAudioPort(enabled ? .speaker : .none)
    }

    /// Binds a native render view to a tile (called by the platform view factory).
    func bindVideoView(_ view: VideoRenderView, tileId: Int) {
        meetingSession?.audioVideo.bindVideoView(videoView: view, tileId: tileId)
    }

    func unbindVideoView(tileId: Int) {
        meetingSession?.audioVideo.unbindVideoView(tileId: tileId)
    }

    // MARK: - Audio session (VoIP)

    private func configureAudioSession(activate: Bool) throws {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, mode: .voiceChat,
                                options: [.allowBluetooth, .allowBluetoothA2DP])
        if activate {
            try session.setActive(true)
        }
    }

    private func deactivateAudioSession() {
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }

    // MARK: - Audio route

    /// Watches the OS audio route (bluetooth headset connecting, wired headphones plugged in,
    /// speaker override) and reports every change, so the in-app UI can mirror the system call
    /// screen's route indicator instead of guessing from its own speaker toggle.
    private func observeAudioRouteChanges(emitInitial: Bool) {
        guard routeObserver == nil else { return }
        routeObserver = NotificationCenter.default.addObserver(
            forName: AVAudioSession.routeChangeNotification, object: nil, queue: .main
        ) { [weak self] _ in
            self?.emitAudioRouteIfChanged()
        }
        // Initial route so the UI is correct before any change (e.g. bluetooth already connected).
        if emitInitial { emitAudioRouteIfChanged() }
    }

    /// The phone-app default the Chime meetings SDK doesn't apply: when a bluetooth or wired
    /// headset is connected at call start, route the call audio to it (unless the user explicitly
    /// forced the speaker). Chime routes via `setPreferredInput`; iOS then uses the paired output.
    private func applyPreferredHeadsetRoute() {
        guard !userForcedSpeaker, let audioVideo = meetingSession?.audioVideo else { return }
        let current = currentAudioRoute()
        guard current != "bluetooth" && current != "headset" else { return }
        let devices = audioVideo.listAudioDevices()
        let preferred = [MediaDeviceType.audioBluetooth, .audioWiredHeadset]
            .compactMap { wanted in devices.first { $0.type == wanted } }.first
        guard let preferred = preferred else { return }
        // Clear any loudspeaker override first — chooseAudioDevice(non-speaker) never does.
        try? AVAudioSession.sharedInstance().overrideOutputAudioPort(.none)
        audioVideo.chooseAudioDevice(mediaDevice: preferred)
    }

    private func stopObservingAudioRouteChanges() {
        if let observer = routeObserver { NotificationCenter.default.removeObserver(observer) }
        routeObserver = nil
        lastRoute = nil
    }

    private func emitAudioRouteIfChanged() {
        let route = currentAudioRoute()
        guard route != lastRoute else { return }
        lastRoute = route
        let outputs = AVAudioSession.sharedInstance().currentRoute.outputs.map { $0.portType.rawValue }
        logger.info(msg: "audio route -> \(route) (outputs: \(outputs))")
        emitter?.emit(["type": "audioRouteChanged", "route": route])
    }

    /// Classifies the session's active output into the event contract's four routes
    /// (`speaker | receiver | bluetooth | headset`).
    private func currentAudioRoute() -> String {
        for output in AVAudioSession.sharedInstance().currentRoute.outputs {
            switch output.portType {
            case .builtInSpeaker, .airPlay:
                return "speaker"
            case .bluetoothA2DP, .bluetoothHFP, .bluetoothLE, .carAudio:
                return "bluetooth"
            case .headphones, .usbAudio:
                return "headset"
            default:
                continue
            }
        }
        return "receiver"
    }

    // MARK: - Helpers

    private func emitState(_ state: String, reason: String? = nil) {
        var event: [String: Any] = ["type": "stateChanged", "state": state]
        if let reason = reason { event["reason"] = reason }
        emitter?.emit(event)
    }
}

// MARK: - AudioVideoObserver

extension ChimeCallManager: AudioVideoObserver {
    func audioSessionDidStartConnecting(reconnecting: Bool) {
        if reconnecting { emitState("reconnecting") }
    }

    func audioSessionDidStart(reconnecting: Bool) {
        emitState("connected")
    }

    func audioSessionDidDrop() {
        emitState("reconnecting")
    }

    func audioSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
        let ok = sessionStatus.statusCode == .ok
        emitState(ok ? "disconnected" : "failed", reason: "\(sessionStatus.statusCode)")
        if !ok {
            emitter?.emit(["type": "error", "code": "sessionEnded",
                           "message": "\(sessionStatus.statusCode)", "fatal": true])
        }
    }

    func audioSessionDidCancelReconnect() {
        emitState("failed", reason: "reconnect cancelled")
    }

    func connectionDidRecover() {
        emitter?.emit(["type": "networkQualityChanged", "quality": "good"])
    }

    func connectionDidBecomePoor() {
        emitter?.emit(["type": "networkQualityChanged", "quality": "poor"])
    }

    func videoSessionDidStartConnecting() {}

    func videoSessionDidStartWithStatus(sessionStatus: MeetingSessionStatus) {}

    func videoSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {}

    func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {}

    func remoteVideoSourcesDidBecomeUnavailable(sources: [RemoteVideoSource]) {}

    func cameraSendAvailabilityDidChange(available: Bool) {}
}

// MARK: - RealtimeObserver

extension ChimeCallManager: RealtimeObserver {
    func attendeesDidJoin(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            emitter?.emit(["type": "participantJoined",
                           "attendeeId": info.attendeeId,
                           "externalUserId": info.externalUserId])
        }
    }

    func attendeesDidLeave(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            emitter?.emit(["type": "participantLeft", "attendeeId": info.attendeeId])
        }
    }

    func attendeesDidDrop(attendeeInfo: [AttendeeInfo]) {
        for info in attendeeInfo {
            emitter?.emit(["type": "participantLeft", "attendeeId": info.attendeeId])
        }
    }

    func attendeesDidMute(attendeeInfo: [AttendeeInfo]) {}
    func attendeesDidUnmute(attendeeInfo: [AttendeeInfo]) {}
    func volumeDidChange(volumeUpdates: [VolumeUpdate]) {}
    func signalStrengthDidChange(signalUpdates: [SignalUpdate]) {}
}

// MARK: - VideoTileObserver

extension ChimeCallManager: VideoTileObserver {
    func videoTileDidAdd(tileState: VideoTileState) {
        if tileState.isLocalTile {
            emitter?.emit(["type": "localVideoAvailable", "tileId": tileState.tileId])
        } else {
            emitter?.emit(["type": "remoteVideoAvailable",
                           "tileId": tileState.tileId,
                           "attendeeId": tileState.attendeeId])
        }
    }

    func videoTileDidRemove(tileState: VideoTileState) {
        meetingSession?.audioVideo.unbindVideoView(tileId: tileState.tileId)
        emitter?.emit(["type": "videoTileRemoved", "tileId": tileState.tileId])
    }

    func videoTileDidPause(tileState: VideoTileState) {}
    func videoTileDidResume(tileState: VideoTileState) {}
    func videoTileSizeDidChange(tileState: VideoTileState) {}
}

File: packages/flutter_amazon_connect_webrtc/android/src/main/kotlin/com/chimeflutter/connect_webrtc/ChimeCallManager.kt
package com.chimeflutter.connect_webrtc

import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.os.Build
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AttendeeInfo
import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.SignalUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.VolumeUpdate
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.RemoteVideoSource
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoRenderView
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import com.amazonaws.services.chime.sdk.meetings.realtime.RealtimeObserver
import com.amazonaws.services.chime.sdk.meetings.session.DefaultMeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSession
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatusCode
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.amazonaws.services.chime.sdk.meetings.utils.logger.LogLevel

/**
 * Owns the Amazon Chime meeting session and coordinates with Android Telecom (via
 * [ConnectTelecomManager]) when system-call UI is requested.
 *
 * Two modes:
 *  - **Telecom** (systemCallUI + API 26+): Telecom owns the call lifecycle, system UI and audio
 *    routing/focus. Chime media is started from the Telecom `onSetActive` callback; this class does
 *    NOT request audio focus or drive `AudioManager` routing.
 *  - **Standalone**: this class requests audio focus, starts a microphone foreground service, and
 *    starts Chime media immediately.
 *
 * Observer callbacks arrive on background threads; [emit] marshals to the main thread (done by the
 * plugin's emitter). NOTE: verify observer signatures against amazon-chime-sdk 0.25.4 when building.
 */
class ChimeCallManager(
    private val context: Context,
    private val emit: (Map<String, Any?>) -> Unit,
) : AudioVideoObserver, RealtimeObserver, VideoTileObserver {

    private val logger = ConsoleLogger(LogLevel.INFO)
    private var meetingSession: MeetingSession? = null
    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    private var telecom: ConnectTelecomManager? = null
    private var usingTelecom = false
    private var started = false
    private var isMuted = false
    private var focusRequest: android.media.AudioFocusRequest? = null

    private val audioVideo get() = meetingSession?.audioVideo

    /** [args] is the full method-channel map: the CallSession fields plus `callKitEnabled`,
     * `callDisplayName`, `callType`. */
    fun join(args: Map<String, Any?>) {
        val configuration = ChimeMeetingSessionAdapter.makeConfiguration(args)
        // An answered incoming (simulated-outbound) call always uses Telecom when available so the
        // system owns audio focus/routing, same as an outgoing system-UI call.
        val asIncoming = args["asIncoming"] as? Boolean ?: false
        val systemCallUI = ((args["callKitEnabled"] as? Boolean ?: false) || asIncoming) &&
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
        val displayName = args["callDisplayName"] as? String ?: "Support"
        val isVideo = (args["callType"] as? String) == "video"

        val session = DefaultMeetingSession(configuration, logger, context)
        meetingSession = session
        session.audioVideo.apply {
            addAudioVideoObserver(this@ChimeCallManager)
            addRealtimeObserver(this@ChimeCallManager)
            addVideoTileObserver(this@ChimeCallManager)
        }
        started = false
        emitState("connecting")

        if (systemCallUI) {
            usingTelecom = true
            val manager = ConnectTelecomManager(context)
            telecom = manager
            if (asIncoming) {
                manager.startAnsweredIncomingCall(
                    displayName = displayName,
                    isVideo = isVideo,
                    onActive = { startMedia() },
                    onDisconnected = { stopMedia() },
                    onMuteChanged = { muted -> applyMuteFromSystem(muted) },
                )
            } else {
                manager.startOutgoingCall(
                    displayName = displayName,
                    isVideo = isVideo,
                    onActive = { startMedia() },
                    onDisconnected = { stopMedia() },
                    onMuteChanged = { muted -> applyMuteFromSystem(muted) },
                )
            }
        } else {
            usingTelecom = false
            requestAudioFocus()
            startMedia()
        }
    }

    /** Starts the Chime media (once). Called immediately (standalone) or from Telecom onSetActive. */
    private fun startMedia() {
        if (started) return
        started = true
        startForegroundService()
        audioVideo?.start()
        audioVideo?.startRemoteVideo()
        if (isMuted) audioVideo?.realtimeLocalMute()
    }

    fun leave() {
        if (usingTelecom) {
            telecom?.endCall() // → onDisconnect → stopMedia()
        } else {
            stopMedia()
        }
    }

    /** Idempotent teardown of the media session. */
    private fun stopMedia() {
        val session = meetingSession ?: return
        session.audioVideo.apply {
            stopLocalVideo()
            stopRemoteVideo()
            removeAudioVideoObserver(this@ChimeCallManager)
            removeRealtimeObserver(this@ChimeCallManager)
            removeVideoTileObserver(this@ChimeCallManager)
            stop()
        }
        meetingSession = null
        started = false
        if (!usingTelecom) abandonAudioFocus()
        stopForegroundService()
        telecom = null
        usingTelecom = false
        emitState("disconnected")
    }

    fun setMuted(muted: Boolean): Boolean {
        val av = audioVideo ?: return false
        val ok = if (muted) av.realtimeLocalMute() else av.realtimeLocalUnmute()
        if (ok) {
            isMuted = muted
            emit(mapOf("type" to "muteChanged", "muted" to muted))
        }
        return ok
    }

    /** Called when the user mutes from the system call UI (Telecom `isMuted` flow). */
    private fun applyMuteFromSystem(muted: Boolean) {
        if (muted != isMuted) setMuted(muted)
    }

    fun setLocalVideoEnabled(enabled: Boolean) {
        val av = audioVideo ?: return
        if (enabled) av.startLocalVideo() else av.stopLocalVideo()
    }

    fun switchCamera() {
        audioVideo?.switchCamera()
    }

    fun setSpeakerphone(enabled: Boolean) {
        if (usingTelecom) {
            // Telecom owns audio routing — route the change through it (never AudioManager directly).
            if (telecom?.setSpeaker(enabled) == true) {
                emit(mapOf("type" to "audioRouteChanged", "route" to if (enabled) "speaker" else "receiver"))
            }
            return
        }
        @Suppress("DEPRECATION")
        audioManager.isSpeakerphoneOn = enabled
        emit(mapOf("type" to "audioRouteChanged", "route" to if (enabled) "speaker" else "receiver"))
    }

    fun bindVideoView(view: VideoRenderView, tileId: Int) {
        audioVideo?.bindVideoView(view, tileId)
    }

    fun unbindVideoView(tileId: Int) {
        audioVideo?.unbindVideoView(tileId)
    }

    // MARK: audio focus (standalone only)

    private fun requestAudioFocus() {
        val request = android.media.AudioFocusRequest
            .Builder(AudioManager.AUDIOFOCUS_GAIN)
            .setAudioAttributes(
                android.media.AudioAttributes.Builder()
                    .setUsage(android.media.AudioAttributes.USAGE_VOICE_COMMUNICATION)
                    .setContentType(android.media.AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build(),
            )
            .build()
        focusRequest = request
        audioManager.requestAudioFocus(request)
        audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
    }

    private fun abandonAudioFocus() {
        focusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
        focusRequest = null
        audioManager.mode = AudioManager.MODE_NORMAL
    }

    // MARK: foreground service

    private fun startForegroundService() {
        val intent = Intent(context, CallForegroundService::class.java)
        context.startForegroundService(intent)
    }

    private fun stopForegroundService() {
        context.stopService(Intent(context, CallForegroundService::class.java))
    }

    private fun emitState(state: String, reason: String? = null) {
        val event = mutableMapOf<String, Any?>("type" to "stateChanged", "state" to state)
        if (reason != null) event["reason"] = reason
        emit(event)
    }

    // MARK: AudioVideoObserver

    override fun onAudioSessionStartedConnecting(reconnecting: Boolean) {
        if (reconnecting) emitState("reconnecting")
    }

    override fun onAudioSessionStarted(reconnecting: Boolean) {
        emitState("connected")
    }

    override fun onAudioSessionDropped() {
        emitState("reconnecting")
    }

    override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
        val ok = sessionStatus.statusCode == MeetingSessionStatusCode.OK
        if (!ok) {
            emit(
                mapOf(
                    "type" to "error", "code" to "sessionEnded",
                    "message" to (sessionStatus.statusCode?.name ?: "unknown"), "fatal" to true,
                ),
            )
        }
        // Route teardown through the same path so the system call UI is dismissed too.
        if (usingTelecom) telecom?.endCall() else stopMedia()
    }

    override fun onAudioSessionCancelledReconnect() {
        emitState("failed", "reconnect cancelled")
    }

    override fun onConnectionRecovered() {
        emit(mapOf("type" to "networkQualityChanged", "quality" to "good"))
    }

    override fun onConnectionBecamePoor() {
        emit(mapOf("type" to "networkQualityChanged", "quality" to "poor"))
    }

    override fun onVideoSessionStartedConnecting() {}
    override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {}
    override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {}
    override fun onRemoteVideoSourceAvailable(sources: List<RemoteVideoSource>) {}
    override fun onRemoteVideoSourceUnavailable(sources: List<RemoteVideoSource>) {}
    override fun onCameraSendAvailabilityUpdated(available: Boolean) {}

    // MARK: RealtimeObserver

    override fun onAttendeesJoined(attendeeInfo: Array<AttendeeInfo>) {
        attendeeInfo.forEach {
            emit(
                mapOf(
                    "type" to "participantJoined",
                    "attendeeId" to it.attendeeId,
                    "externalUserId" to it.externalUserId,
                ),
            )
        }
    }

    override fun onAttendeesLeft(attendeeInfo: Array<AttendeeInfo>) {
        attendeeInfo.forEach { emit(mapOf("type" to "participantLeft", "attendeeId" to it.attendeeId)) }
    }

    override fun onAttendeesDropped(attendeeInfo: Array<AttendeeInfo>) {
        attendeeInfo.forEach { emit(mapOf("type" to "participantLeft", "attendeeId" to it.attendeeId)) }
    }

    override fun onAttendeesMuted(attendeeInfo: Array<AttendeeInfo>) {}
    override fun onAttendeesUnmuted(attendeeInfo: Array<AttendeeInfo>) {}
    override fun onVolumeChanged(volumeUpdates: Array<VolumeUpdate>) {}
    override fun onSignalStrengthChanged(signalUpdates: Array<SignalUpdate>) {}

    // MARK: VideoTileObserver

    override fun onVideoTileAdded(tileState: VideoTileState) {
        if (tileState.isLocalTile) {
            emit(mapOf("type" to "localVideoAvailable", "tileId" to tileState.tileId))
        } else {
            emit(
                mapOf(
                    "type" to "remoteVideoAvailable",
                    "tileId" to tileState.tileId,
                    "attendeeId" to tileState.attendeeId,
                ),
            )
        }
    }

    override fun onVideoTileRemoved(tileState: VideoTileState) {
        audioVideo?.unbindVideoView(tileState.tileId)
        emit(mapOf("type" to "videoTileRemoved", "tileId" to tileState.tileId))
    }

    override fun onVideoTilePaused(tileState: VideoTileState) {}
    override fun onVideoTileResumed(tileState: VideoTileState) {}
    override fun onVideoTileSizeChanged(tileState: VideoTileState) {}
}
